lib/clibuddy/command.rb

```diff
@@ -0,0 +1,65 @@
+module CLIBuddy
+  class Command
+
+    attr_reader :name
+    attr_accessor :flow, :definition, :usage
+
+    def initialize(name)
+      @name = name
+    end
+
+    def required_args
+      return [] if definition.arguments.empty?
+      definition.arguments.select { |a| !a.optional? }
+    end
+
+    def optional_args
+      return [] if definition.arguments.empty?
+      definition.arguments.select { |a| a.optional? }
+    end
+
+    # TODO This should probably be done in the formatter, or somewhere else (maybe>)
+    # when user types `--help` then we show the full usage text plus usage string,
+    # if they typo a command/used unrecognized arguments we show the short usage test plus usage string.
+    def full_usage
+      u = usage[:full].join("\n")
+      u += "\n\nUsage: #{name} "
+      u += required_args.map { |a| a.name }.join(" ")
+      u += " [options...]\n    "
+      oa = optional_args.map { |a| a.short_name + ", " + a.long_name.ljust(30) + a.description.join(" ") }
+      u += oa.join("\n    ")
+    end
+
+  end
+
+  CommandDefinition = Struct.new(:arguments)
+
+  class CommandDefinitionArg
```

Comment:
I was initially thinking `class CommandDefinitionArg < Struct.new(:name, :param, :description)`. This gives you the same initialize signature and saves you the `attr_reader` and `#initialize`.